pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -416,15 +415,18 @@ private void buildDedupMeta(ImmutableSegmentImpl immutableSegment) {
 
   private void handleUpsert(ImmutableSegmentImpl immutableSegment) {
     String segmentName = immutableSegment.getSegmentName();
-    Integer partitionGroupId =
-        SegmentUtils.getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager,
-            _primaryKeyColumns.get(0));
-    Preconditions.checkNotNull(partitionGroupId,
-        String.format("PartitionGroupId is not available for segment: '%s' (upsert-enabled table: %s)", segmentName,
+    Integer partitionGroupId = SegmentUtils
+        .getRealtimeSegmentPartitionId(segmentName, _tableNameWithType, _helixManager, _primaryKeyColumns.get(0));
+    Preconditions.checkNotNull(partitionGroupId, String
+        .format("PartitionGroupId is not available for segment: '%s' (upsert-enabled table: %s)", segmentName,
             _tableNameWithType));
     PartitionUpsertMetadataManager partitionUpsertMetadataManager =
         _tableUpsertMetadataManager.getOrCreatePartitionManager(partitionGroupId);
+
     ThreadSafeMutableRoaringBitmap validDocIds = new ThreadSafeMutableRoaringBitmap();
+    if (immutableSegment.getValidDocIds() != null) {
```

Comment:
I don't think this is the correct way to handle it. We should get the persisted valid doc id, and iterate over it to get the `recordInfoIterator`

Reply:
Gotcha, thanks for review.
For immutableSegmentImpl, it already has `validDocsIds` variable for keeping in memory states, I was think of loading the validDocsSnapshots from disk to `validDocsIds` variable directly. Is there any concern/risk on this?
I can also add another variable call validDocsSnapshots that loaded from disk. Iterate over validDocsSnapshots to get the recordInfoInteractor and re-generate validDocIds in PartitionUpsertMetadataManager (addSegment).

Reply:
fixed [here](https://github.com/apache/pinot/pull/9062/commits/56fb2a6b629473f5a64eb945597092854a39c3ba)